### ISSUE TITLE
Provide more ways to interaction with Instant

### ISSUE THREAD

**Issue by dignifiedquire** [communityvi/async-time-mock#19]
Would be great to have the following to better integrate into code

- Conversion to `std::time::Instant`
- Conversion to `tokio::time::Instant`
- Derefs?

**Comment by FSMaxB** [comment]
### `async_time_mock_core::Instant`
Do you mean you want to convert `async_time_mock_core::Instant` into `std::time::Instant` or `tokio::time::Instant`? https://github.com/communityvi/async-time-mock/blob/3f389d8f2b7db1bfad38effef1f2ba800829ff80/async-time-mock-core/src/instant.rs#L4-L8

If so, this is not possible because a mock instant is completely different from a real instant. You can't convert it from one to the other because the rate at which real time passes is decoupled from the rate that the test time passes.

What you **can** do is to take the duration between two instants and apply that to a real instant though.

### `async_time_mock_tokio::Instant`

Or do you mean to get the instant out of something like `async_time_mock_tokio::Instant`?
https://github.com/communityvi/async-time-mock/blob/3f389d8f2b7db1bfad38effef1f2ba800829ff80/async-time-mock-tokio/src/instant.rs#L5-L10

You can match on it like so:
```rust
match my_instant {
    Real(instant) => Some(instant),
    _ => None,
}
```

But you obviously aren't guaranteed to get a real instant, because you can only get one if it is not mocked.

Implementing `Deref` is not possible for the same reason `Option<T>` doesn't implement `Deref<T>`: There may or may not be valid content, so you can't unconditionally provide a reference to it.

What I could do though is something like the following:
```rust
pub fn real(&self) -> Option<tokio::time::Instant> {
    match self {
        Real(instant) => Some(instant),
        _ => None,
    }
}
```

Then you could do something like:
```rust
if let Some(real_instant) = my_instant {
  // do something with it that only happens when time is not mocked
}
```
That's not too much different from doing the match manually as described above though.

Whether having code that behaves differently with real and mocked time is a good idea is a different question.

**Comment by dignifiedquire** [comment]
thanks for the detailed answer, managed to work around this actually. The one thing I was still missing is Instant::elapsed() which is a nice helper I often use, which I think should be possible.

**Comment by FSMaxB** [comment]
> The one thing I was still missing is Instant::elapsed() which is a nice helper I often use, which I think should be possible.

There's two ways `elapsed` could be implemented:
1. Storing a reference to the `TimerRegistry` in the instant. This doesn't sound like a good idea because that would make it impossible to derive `Copy` on it (so you would need `.clone()` everywhere, which makes it much harder to use)
2. Require a `MockableClock` parameter like `fn elapsed(&self, &MockableClock) -> Duration`

1 could be worked around by storing all timer registries in some kind of global registry, and refer to them by index, but I'd rather not do that.

2 is certainly feasible!

**Comment by FSMaxB** [comment]
I will add `elapsed` once I find the time (which I'm not sure yet when that will be).

**Comment by FSMaxB** [comment]
`Instant::elapsed` is now released in `0.1.0`

**Comment by dignifiedquire** [comment]
thanks!
